src/core/api/api_document.pl

```diff
@@ -164,14 +163,53 @@
         do_or_die(insert_document(Transaction, Document, Captures_In, Id, _Dependencies, Captures_Out),
                   error(document_insertion_failed_unexpectedly(Document), _))).
 
-prepare_full_replace(true, schema, Stream, Transaction, Tail_Stream) :-
+api_insert_document_unsafe_(schema, Transaction, Document, state(Captures), Id, Captures) :-
+    do_or_die(
+        insert_schema_document_unsafe(Transaction, Document),
+        error(document_insertion_failed_unexpectedly(Document), _)),
+    do_or_die(
+        Id = (Document.get('@id')),
+        error(document_has_no_id_somehow, _)).
+api_insert_document_unsafe_(instance, Transaction, Document, state(Captures_In), Id, Captures_Out) :-
+    do_or_die(
+        insert_document_unsafe(Transaction, Document, Captures_In, Id, Captures_Out),
+        error(document_insertion_failed_unexpectedly(Document), _)).
+
+insert_documents_(true, Graph_Type, Stream, Transaction, Captures_Var, Ids) :-
```

Comment:
This looks good. Note though that it might be slightly better to use first-argument indexing with a discriminator (in this case true/false) than testing a first argument, and using a cut.

Reply:
https://github.com/terminusdb/terminusdb/pull/1117